api/agent/lb_hash.go

```diff
@@ -0,0 +1,74 @@
+/* The consistent hash ring from the original fnlb.
+   The behaviour of this depends on changes to the runner list leaving it relatively stable.
+*/
+
+package agent
+
+import (
+	"context"
+	"time"
+
+	"github.com/dchest/siphash"
+	"github.com/fnproject/fn/api/models"
+	"github.com/sirupsen/logrus"
+)
+
+type chPlacer struct {
+}
+
+func NewCHPlacer() Placer {
+	return &chPlacer{}
+}
+
+// This borrows the CH placement algorithm from the original FNLB.
+// Because we ask a runner to accept load (queuing on the LB rather than on the nodes), we don't use
+// the LB_WAIT to drive placement decisions: runners only accept work if they have the capacity for it.
+func (p *chPlacer) PlaceCall(np NodePool, ctx context.Context, call *call, lbGroupID string) error {
+	deadline := call.slotDeadline
+
+	// The key is just the path in this case
+	key := call.Path
+	sum64 := siphash.Hash(0, 0x4c617279426f6174, []byte(key))
+
+	for {
+		if time.Now().After(deadline) {
+			return models.ErrCallTimeoutServerBusy
+		}
+
+		runners := np.Runners(lbGroupID)
+		if len(runners) < 1 {
+			time.Sleep(retryWaitInterval)
```

Comment:
should be sleep(min(retryWaitInterval , delta(now, deadline)))

Reply:
But better to select on both, since agent Submit could cancel this (eg. shutdown)